HLTrigger/Timer/plugins/ThroughputService.cc

```diff
@@ -68,19 +90,63 @@ void ThroughputService::preGlobalBeginRun(edm::GlobalContext const& gc) {
     // book MonitorElement's for this run
     edm::Service<DQMStore>()->meBookerGetter(bookTransactionCallback);
   } else {
-    std::cerr << "No DQMStore service, aborting." << std::endl;
-    abort();
+    m_sourced_events = nullptr;
+    m_retired_events = nullptr;
   }
 }
 
 void ThroughputService::preSourceEvent(edm::StreamID sid) {
-  auto timestamp = std::chrono::steady_clock::now();
-  m_sourced_events->Fill(std::chrono::duration_cast<std::chrono::duration<double>>(timestamp - m_startup).count());
+  auto timestamp = std::chrono::high_resolution_clock::now();
+  auto interval = std::chrono::duration_cast<std::chrono::duration<double>>(timestamp - m_startup).count();
+  if (m_enable_dqm) {
+    m_sourced_events->Fill(interval);
+  }
 }
 
 void ThroughputService::postEvent(edm::StreamContext const& sc) {
-  auto timestamp = std::chrono::steady_clock::now();
-  m_retired_events->Fill(std::chrono::duration_cast<std::chrono::duration<double>>(timestamp - m_startup).count());
+  auto timestamp = std::chrono::high_resolution_clock::now();
+  auto interval = std::chrono::duration_cast<std::chrono::duration<double>>(timestamp - m_startup).count();
+  if (m_enable_dqm) {
+    m_retired_events->Fill(interval);
+  }
+  ++m_counter;
+  if (m_counter % m_resolution == 0) {
+    m_events.push_back(timestamp);
+  }
+}
+
+void ThroughputService::postEndJob() {
+  if (m_counter < 2 * m_resolution) {
+    // not enough mesurements to estimate the throughput
+    edm::LogWarning("ThroughputService") << "Not enough events to measure the throughput with a resolution of "
+                                         << m_resolution << " events";
+    return;
+  }
+
+  edm::LogInfo info("ThroughputService");
+
+  if (m_print_event_summary) {
+    for (uint32_t i = 0; i < m_events.size(); ++i) {
+      info << std::setw(8) << (i + 1) * m_resolution << ", "
+           << std::setprecision(6) << edm::TimeOfDay(std::chrono::duration_cast<std::chrono::microseconds>(m_events[i].time_since_epoch())) << "\n";
```

Comment:
this results in an output like
```
      10, 19-Mar-2020 08:09:54.491249 CET
00000020, 19-Mar-2020 08:09:59.947239 CET
00000030, 19-Mar-2020 08:10:05.350185 CET
00000040, 19-Mar-2020 08:10:10.558139 CET
00000050, 19-Mar-2020 08:10:15.448258 CET
00000060, 19-Mar-2020 08:10:20.013977 CET
00000070, 19-Mar-2020 08:10:24.487986 CET
00000080, 19-Mar-2020 08:10:29.889229 CET
00000090, 19-Mar-2020 08:10:34.753967 CET
00000100, 19-Mar-2020 08:10:39.220913 CET
```

Looks like `TimeOfDay` used together with `LogInfo` does not properly reset the stream flags ?

Reply:
I'm actually confused as to how this works at all: `LogInfo` does not inherit from `ostream`, so how does
```
std::ostream& operator<<(std::ostream& os, TimeOfDay const& tod)
```
even work ?

Reply:
`LogInfo::operator<<()`
https://github.com/cms-sw/cmssw/blob/4fa6d2a97df1f2b18fceeaa6e093975311570c5f/FWCore/MessageLogger/interface/MessageLogger.h#L264-L269
forwards the call to `MessageSender::operator<<()`
https://github.com/cms-sw/cmssw/blob/4fa6d2a97df1f2b18fceeaa6e093975311570c5f/FWCore/MessageLogger/interface/MessageSender.h#L38-L43
who further forwards the call to `ErrorObj::operator<<()`
https://github.com/cms-sw/cmssw/blob/4fa6d2a97df1f2b18fceeaa6e093975311570c5f/FWCore/MessageLogger/interface/ErrorObj.icc#L41-L44
who forwards the call to `ErrorObj::opltlt()`
https://github.com/cms-sw/cmssw/blob/4fa6d2a97df1f2b18fceeaa6e093975311570c5f/FWCore/MessageLogger/interface/ErrorObj.icc#L23-L26
who finally forwards the call to `std::ostringstream::operator<<()`.

Reply:
... I see, thanks !